### ISSUE TITLE
Unreliable Connection

### ISSUE THREAD

**Issue by FlantasticDan** [clvLabs/PyATEMMax#12]
I'm so impressed with this module and the functionality you've been able to reverse engineer, thank you so much for making this available.

I'm running into an issue that persists across multiple ATEM devices and operating systems.  Sometimes the code connects to the ATEM and sometimes it just endlessly tries (but fails) to connect.  My switchers all run firmware `8.1.1` so I understand there may be compatibility problems but the intermittent nature of this issue suggests something more to me.

I've generated some logs that I hope will help troubleshoot using this test code:
```python
import logging

import PyATEMMax

logging.basicConfig(filename='pyatemmax.log', encoding='utf-8', level=logging.DEBUG)

switcher = PyATEMMax.ATEMMax()
switcher.setSocketLogLevel(logging.DEBUG)

switcher.connect("192.168.2.200")
switcher.waitForConnection()
logging.log(logging.INFO, 'Connected')
```

Attached are a couple sets of logs (from different module versions), one where it successfully connects, and one where it continuously fails to connect:

**v1.0b4**
[pyatemmax-connected.log](https://github.com/clvLabs/PyATEMMax/files/7805539/pyatemmax-connected.log)
[pyatemmax-failed.log](https://github.com/clvLabs/PyATEMMax/files/7805540/pyatemmax-failed.log)

**1.0b6**
[pyatemmax-connected-2.log](https://github.com/clvLabs/PyATEMMax/files/7805560/pyatemmax-connected-2.log)
[pyatemmax-failed-2.log](https://github.com/clvLabs/PyATEMMax/files/7805561/pyatemmax-failed-2.log)

Thanks again, if there's anything I can do to help please let me know!

**Comment by clvLabs** [comment]
Hi @FlantasticDan !

First of all, thanks for the detailed and complete issue report.

As you point out, the intermitent nature of your problem seems to discard the compatibility problems possibility.

Yesterday night I took a quick look at the logs you sent. It's curious to see that in the _failed_ cases, the library could phisically (TCP) connect to the device and start a conversation with it. But, at a certain point, where the _connected_ cases assume the initial data snapshot has been received and mark the connection as established, the _failed_ cases continue to transmit messages back and forth, as if the snapshot was not complete.

I have no clear hypothesis abouth the source of the problem, I'll take a deeper look at the hex content of the messages to see if I can find something. Please consider this can be a tedious task... it can take some time.

I'll get back to you if I need your help or when I've found the issue.

**Comment by clvLabs** [comment]
Hello again @FlantasticDan,

Think I found something...

In `ATEMConnectionManager.py`, inside `_runLoop()` there's this comment from the original Arduino library code:

```python
# If a packet is 12 bytes long it indicates that all the initial information
# has been delivered from the ATEM and we can begin to answer back on every request
# Currently we don't know any other way to decide if an answer should be sent back...
# The QT lib uses the "InCm" command to indicate this, but in the latest version of the firmware (2.14)
# all the camera control information comes AFTER this command, so it's not a clear ending token anymore.
# However, I'm not sure if I checked the lastRemotePacketID of the packets with the additional camera control info - if it was a resend,
# "InCm" may still indicate the number of the last init-packet and that's all I need to request the missing ones....
```

Looking at the logs you sent, both of the `connected` samples had 12-byte messages on them while the `failed` ones didn't.

This makes the code think the initial snapshot is not complete and the connection is not fully established.

The first thing I'd like to ask you is to try changing the
```python
switcher.waitForConnection()
```
for a 
```python
switcher.waitForConnection(waitForFullHandshake=False)
time.sleep(5)
```
in your example code and try it.

With this, the wait stops as soon as the `UDP` connection is confirmed and we add an extra 5 seconds of wait time to allow the full snapshot to be sent.

This should do the trick, but I don't really like it... so I prepared a `test` version for you to try.
[pyatemmax-issue-12-fix-test.zip](https://github.com/clvLabs/PyATEMMax/files/7885333/pyatemmax-issue-12-fix-test.zip)

Reading the logs I confirmed that the `camera control` messages mentioned in the comment above coming after the `InCm` message are actually resends, so I've tried assuming the snapshot is finished when the `InCm` message is received.

Unzip the attached file in the same folder where your sample code is. This will create a `PyATEMMax` folder with the modified code.

Please tell me how it went...

**Comment by clvLabs** [comment]
Any news, could you try it ?

**Comment by FlantasticDan** [comment]
@clvLabs Sorry about disappearing, I've had to shift my focus away from ATEM products as of late but will be getting back to them in the next couple of weeks.  Thanks for working on this, I will test this next I have some time with an ATEM.

**Comment by clvLabs** [comment]
I've already seem the `InCm` command interpreted as `Initialization Complete` in some other ATEM libs, so I'm opening a branch to make the change and accept `InCm` as the init sequence completer.

**Comment by clvLabs** [comment]
Sorry... I think it's time I go thru some tutorials on how to handle issues and branches in github 😅 

I created a branch named ' 12-unreliable-connection` from this same issue, committed the changes on that branch and later merged in into master with [this commit](https://github.com/clvLabs/PyATEMMax/commit/3946bd4470528e91f8775a5e10f0e5867ebbe084).

Next time I hope I can get github to automatically show notifications and links to the commits in the issue.

**Comment by clvLabs** [comment]
The change is available in version [1.0b9](https://clvlabs.github.io/PyATEMMax/pyatemmax/update/2022/09/16/release-1-0-b9.html)

**Comment by luizdemario** [comment]
do que se tratam erros :ATEMMax object has no attribute data, geralmente em controles de saturação,, ou lifts rgb ?
